otx/core/data/base_dataset_adapter.py

```diff
@@ -0,0 +1,177 @@
+                val_data_type = self._select_data_type(val_data_candidates)
+                self.dataset[Subset.VALIDATION] = DatumaroDataset.import_from(val_data_roots, format=val_data_type)
+
+            if Subset.VALIDATION not in self.dataset:
+                # TODO: auto_split
+                pass
+
+        if test_data_roots:
+            test_data_candidates = self._detect_dataset_format(path=test_data_roots)
+            test_data_type = self._select_data_type(test_data_candidates)
+            self.dataset[Subset.TESTING] = DatumaroDataset.import_from(test_data_roots, format=test_data_type)
+            self.is_train_phase = False
+
+        if unlabeled_data_roots is not None:
+            self.dataset[Subset.UNLABELED] = DatumaroDataset.import_from(unlabeled_data_roots, format="image_dir")
+
+        return self.dataset
+
+    @abstractmethod
+    def convert_to_otx_format(self, datumaro_dataset: dict) -> Tuple[DatasetEntity, LabelSchemaEntity]:
+        """Convert DatumaroDataset to the DatasetEntity.
+
+        Args:
+            datumaro_dataset (dict): A Dictionary that includes subset dataset(DatasetEntity)
+
+        Returns:
+            DatasetEntity:
+        """
+        raise NotImplementedError
+
+    def _detect_dataset_format(self, path: str) -> str:
+        """Detect dataset format (ImageNet, COCO, ...)."""
+        return datumaro.Environment().detect_dataset(path=path)
+
+    def _generate_empty_label_entity(self) -> LabelGroup:
+        """Generate Empty Label Group for H-label, Multi-label Classification."""
+        empty_label = LabelEntity(name="Empty label", is_empty=True, domain=self.domain)
+        empty_group = LabelGroup(name="empty", labels=[empty_label], group_type=LabelGroupType.EMPTY_LABEL)
+        return empty_group
+
+    def _generate_default_label_schema(self, label_entities: list) -> LabelSchemaEntity:
+        """Generate Default Label Schema for Multi-class Classification, Detecion, Etc."""
+        label_schema = LabelSchemaEntity()
+        main_group = LabelGroup(
+            name="labels",
+            labels=label_entities,
+            group_type=LabelGroupType.EXCLUSIVE,
+        )
+        label_schema.add_group(main_group)
+        return label_schema
+
+    def _prepare_label_information(
+        self,
+        datumaro_dataset: dict,
+    ) -> dict:
+        # Get datumaro category information
+        if self.is_train_phase:
+            label_categories_list = (
+                datumaro_dataset[Subset.TRAINING].categories().get(DatumaroAnnotationType.label, None)
+            )
+        else:
+            label_categories_list = (
+                datumaro_dataset[Subset.TESTING].categories().get(DatumaroAnnotationType.label, None)
+            )
+        category_items = label_categories_list.items
+
+        # Get the 'label_groups' information
+        if hasattr(label_categories_list, "label_groups"):
```

Comment:
Hasattr pattern should be avoided in high level code. It's usage means that there is a mistake in design and the LSP is broken somewhere in the inheritance chain. Perhaps we can replace it with None check here